File: targets/java-web/src/main/java/com/example/vulnapp/vulns/Vuln24OpenRedirect.java
package com.example.vulnapp.vulns;

import org.springframework.http.*;
import org.springframework.web.bind.annotation.*;
import java.net.URI;

@RestController
public class Vuln24OpenRedirect {
    @GetMapping("/vuln24/redirect")
    public ResponseEntity<Void> redirect(@RequestParam("url") String url) {
        // VULN:VULN-24:CWE-601:taint open redirect — unvalidated redirect target
        HttpHeaders headers = new HttpHeaders();
        headers.setLocation(URI.create(url));
        return new ResponseEntity<>(headers, HttpStatus.FOUND);
    }
}

File: targets/kotlin-web/src/main/kotlin/com/example/vulnapp/vulns/Vuln24OpenRedirect.kt
package com.example.vulnapp.vulns

import org.springframework.web.bind.annotation.*


@RestController
class Vuln24OpenRedirect {

    @GetMapping("/vuln24/redirect")
    fun redirect(@RequestParam url: String): org.springframework.http.ResponseEntity<Void> {
        // VULN:VULN-24:CWE-601:taint open redirect — unvalidated redirect target
        val h = org.springframework.http.HttpHeaders(); h.location = java.net.URI.create(url)
        return org.springframework.http.ResponseEntity(h, org.springframework.http.HttpStatus.FOUND)
    }
}
